README.md
# [shunt](https://github.com/odb/shunt)

##### Simple Shell Testing Pseudo-Framework

[![shunt](https://raw.github.com/odb/shunt/master/screenshots/shunt.png)](https://raw.github.com/odb/shunt/master/screenshots/shunt.png)

### What!? Why?

For a recent project, I wanted a very simple way to run some simple tests against my script output.
I know there are a number of these kinds of things out there, but most of the ones I looked at were more complex then I wanted.
To start, I created a basic shell script to test my scripts, but as I began to add to it, and want more from it, I decided to take
a few minutes and build a pseudo-framework out of it.

I call it a "pseudo-framework" because it's really more of a helper, which gives you a handfull of assertions to run against
bash commands. Well, that and it doesn't really have a name.

### How?

##### Install

Note, in shunt `latest` means latest stable version.

    curl -L https://raw.github.com/odb/shunt/master/install.sh | bash
    # installs latest to ~/.bin/shunt

    curl -L https://raw.github.com/odb/shunt/master/install.sh | bash -s master
    # installs master to ~/.bin/shunt

    curl -L https://raw.github.com/odb/shunt/master/install.sh | bash -s global
    # installs latest to /usr/local/bin/shunt

    curl -L https://raw.github.com/odb/shunt/master/install.sh | bash -s master local
    # installs master to ./shunt.sh

##### Basic Usage

A basic test file looks like this:

    # file: tests.sh

    function run_tests {
    ####################################################
    # Tests go here.
    ####################################################
      COMMAND="/path/to/your/command"
      assert_grep "$COMMAND" "Usage" \
        "deplay usage without params"
      assert_grep "$COMMAND --help" "Usage" \
        "deplay usage with help"
      assert_grep "$COMMAND --arg2 foobar" "Usage" \
        "deplay usage without required arg"
      refute_grep "$COMMAND --arg1 foobar" "Usage" \
        "work with required arg"
    ####################################################
    }

Run like this:

    $ ./shunt.sh ./tests.sh

Usage:
``` bash
Usage: ./shunt.sh <test files>

Options:
--plain    Disable colors and icons.
--quiet    Do not print error messages.
--verbose  Display success messages.
--version  Display version information.
--help     Display this message.
```

> See `test/testOne.sh` for more examples.

##### Before / After Hooks

In addition to assertion, shunt also supports before and after hooks. Simply define a `before` or `after` function.

    function before {
        ./some_setup_script.sh
        echo "Running before shunt assertions.
    }

    function after {
        ./some_cleanup_script.sh
        echo "Running after shunt assertions.
    }

##### Assertions

Here's a full list of assertions at the time of this writing:

* `assert "CMD" "FAIL MESSAGE"`
* `refute "CMD" "FAIL MESSAGE"`
* `assert_equal "FIRST" "SECOND" "FAIL MESSAGE"`
* `refute_equal "FIRST" "SECOND" "FAIL MESSAGE"`
* `assert_numeq "FIRST" "SECOND" "FAIL MESSAGE"`
* `refute_numeq "FIRST" "SECOND" "FAIL MESSAGE"`
* `assert_grep "CMD" "GREP" "FAIL MESSAGE"`
* `refute_grep "CMD" "GREP" "FAIL MESSAGE"`
* `assert_file "FILE" "FAIL MESSAGE"`
* `refute_file "FILE" "FAIL MESSAGE"`
* `assert_dir "DIR" "FAIL MESSAGE"`
* `refute_dir "DIR" "FAIL MESSAGE"`


### Development

* To run tests use: `make test`
* To update shml.sh, use: `make shml`

### Additional Screenshots

[![shunt --verbose](https://raw.github.com/odb/shunt/master/screenshots/shunt_verbose.png)](https://raw.github.com/odb/shunt/master/screenshots/shunt_verbose.png)

[![shunt --quiet](https://raw.github.com/odb/shunt/master/screenshots/shunt_quiet.png)](https://raw.github.com/odb/shunt/master/screenshots/shunt_quiet.png)

[![shunt --plain](https://raw.github.com/odb/shunt/master/screenshots/shunt_plain.png)](https://raw.github.com/odb/shunt/master/screenshots/shunt_plain.png)



CHANGELOG.md
## CHANGELOG

### 0.2.3

* Fixing multi-file run errors for issues #11.

### 0.2.2

* Update assert|refute_grep error messaging.

### 0.2.1

* Adding line break after Running message (Issue: #12)
* Tweaking assert|refute_grep (Issue: #13)
* Fixing refute error message.

### 0.2.0

* Renaming to 'shunt'.
* Adding CLIstyle for coloring and icons.
* Depricating legacy method of sourcing shunt.
* Adding error trapping and output. Disabled via quite flag.
* Adding verbose test output via verbose flag.
* Fixing issues with before and after show errors when not present. See Issue #5.

### 0.1.1

* Obfuscating internal functions and variables.
* Adding installer (see README).
* Adding argument support for CLIunit.sh (see: '--help')

### 0.1.0

* Supporting ./CLIunit.sh ./test.sh method of running. See Issue #2.

### 0.0.3

* Adding before and after handling. See README for details.

### 0.0.2

* Adding color (thanks [jdorfman](https://github.com/jdorfman)).
* Fixing critical issues with some refute tests.
* Adding `example.sh` for more examples. It also doubles as a sort of test suite.

### 0.0.1

* Initial migration from gist.


test/test.sh
#!/usr/bin/env bash

./CLIunit.sh ./test/testOne.sh ./test/testTwo.sh

# vim: ft=sh:


test/legacy.sh
#!/usr/bin/env bash

# Setup
##
touch /tmp/cliunit.file
mkdir -p /tmp/cliunit.dir

# Test Examples
# -------------
#
# NOTE: This is the legacy way of using 'CLIunit' (0.0.x version)
# whereby CLIunit is sourced at the end of the suite.
#
# I'm keeping this functionality, for backwards compatability and
# certain uses cases where it make sense (they may exist).
#
# The functionality may be depricated in the future.
##

before_check=false
function before {
  before_check=true
}

function after {
  echo "If you see this, after worked."
}

function run_tests {

  assert "true"  "should assert truth"
  refute "false" "should refute truth"

  assert_equal "true" "true" "should assert equality"
  refute_equal "true" "false" "should refute equality"

  assert_numeq "1111" "1111" "should assert numerical equality"
  refute_numeq "1111" "1112" "should refute numerical equality"

  assert_grep "echo foobar" "foobar" "should assert via grep"
  refute_grep "echo foobar" "notfoobar" "should refute via grep"

  assert_file "/tmp/cliunit.file" "should assert file existence"
  refute_file "/tmp/cliunit.bad"  "should refute file existence"

  assert_dir "/tmp/cliunit.dir" "should assert directory existence"
  refute_dir "/tmp/cliunit.bad" "should refute directory existence"

  sleep 0.5 # showing Duration > 0

  assert "false" "should fail, to show failure output"

  assert "$before_check" "before failed"
}
source ./CLIunit.sh

# vim: ft=sh:


test/testOne.sh
#!/usr/bin/env bash

# Setup
##
touch /tmp/shunt.file
mkdir -p /tmp/shunt.dir

# Test Examples
##

tmp_file=/tmp/shunt.file
tmp_dir=/tmp/shunt.dir
function before {
  touch $tmp_file
  mkdir -p $tmp_dir
}

function after {
  echo -e "\n\nIf you see this, after worked."
  rm -rf $tmp_file $tmp_dir
}

function run_tests {
  assert "true"  "should assert truth"
  refute "false" "should refute truth"

  assert_equal "true" "true" "should assert equality"
  refute_equal "true" "false" "should refute equality"

  assert_numeq "1111" "1111" "should assert numerical equality"
  refute_numeq "1111" "1112" "should refute numerical equality"

  assert_grep "echo foobar" "foobar" "should assert via grep"
  refute_grep "echo foobar" "notfoobar" "should refute via grep"

  assert_file "/tmp/shunt.file" "should assert file existence"
  refute_file "/tmp/shunt.bad"  "should refute file existence"

  assert_dir "/tmp/shunt.dir" "should assert directory existence"
  refute_dir "/tmp/shunt.bad" "should refute directory existence"

  # test before
  sleep 0.5 # showing Duration > 0
  assert_file "$tmp_file" "before or assert_file didn't work"
  assert "cat missing.file" "should fail, to show failure output"
}

# vim: ft=sh:


shunt.sh
#!/usr/bin/env bash
################################################################################
#
# Simple Shell Testing Psudo-Framework
# ----------------------------------
#
# Source: https://github.com/odb/shunt
# Author: Joshua Mervine (@mervinej)
#
# ----------------------------------
# Contributors:
# ----------------------------------
# - Justin Dorfman (@jdorfman)
#
################################################################################

SHUNT_VERSION="0.2.4"

# Including an
# Update version with `make shml`
#SHML:START
#************************************************#
#        shml - Shell Markup Framework
#                    Beta
#        by Justin Dorfman - @jdorfman
#        && Joshua Mervine - @mervinej
#
#               Inspired by:
#
#      Kiyor Cia, Jeff Foard FLOZz' MISC,
#           Mark Otto & Dave Gandy
#
#          https://github.com/odb/shml
#************************************************#

# Foreground (Text)
##
function color {
  local __end='\033[39m'
  local __color=$__end # end by default
  case "$1" in
    end|off|reset)       __color=$__end;;
    black|000000)        __color='\033[30m';;
    red|F00BAF)          __color='\033[31m';;
    green|00CD00)        __color='\033[32m';;
    yellow|CDCD00)       __color='\033[33m';;
    blue|0286fe)         __color='\033[34m';;
    magenta|e100cc)      __color='\033[35m';;
    cyan|00d3cf)         __color='\033[36m';;
    gray|e4e4e4)         __color='\033[90m';;
    darkgray|4c4c4c)     __color='\033[91m';;
    lightgreen|00fe00)   __color='\033[92m';;
    lightyellow|f8fe00)  __color='\033[93m';;
    lightblue|3a80b5)    __color='\033[94m';;
    lightmagenta|fe00fe) __color='\033[95m';;
    lightcyan|00fefe)    __color='\033[96m';;
    white|ffffff)        __color='\033[97m';;
  esac
  if test "$2"; then
    echo -en "$__color$2$__end"
  else
    echo -en "$__color"
  fi
}
function c {
  color "$@"
}

# Background
##
function background {
  local __end='\033[49m'
  local __color=$__end # end by default
  case "$1" in
    end|off|reset)       __color=$__end;;
    black|000000)        __color='\033[40m';;
    red|F00BAF)          __color='\033[41m';;
    green|00CD00)        __color='\033[42m';;
    yellow|CDCD00)       __color='\033[43m';;
    blue|0286fe)         __color='\033[44m';;
    magenta|e100cc)      __color='\033[45m';;
    cyan|00d3cf)         __color='\033[46m';;
    gray|e4e4e4)         __color='\033[47m';;
    darkgray|4c4c4c)     __color='\033[100m';;
    lightred)            __color='\033[101m';;
    lightgreen|00fe00)   __color='\033[102m';;
    lightyellow|f8fe00)  __color='\033[103m';;
    lightblue|3a80b5)    __color='\033[104m';;
    lightmagenta|fe00fe) __color='\033[105m';;
    lightcyan|00fefe)    __color='\033[106m';;
    white|fffff)         __color='\033[107m';;
  esac

  if test "$2"; then
    echo -en "$__color$2$__end"
  else
    echo -en "$__color"
  fi
}
function bg {
  background "$@"
}

## Color Bar
function color-bar {
  if test "$2"; then
    for i in "$@"; do
      echo -en "$(background "$i" " ")"
    done; echo
  else
    for i in {16..21}{21..16}; do
      echo -en "\033[48;5;${i}m \033[0m"
    done; echo
  fi
}
function bar {
  color-bar "$@"
}

## Attributes
##
function attribute {
  local __end='\033[0m'
  local __attr=$__end # end by default
  case "$1" in
    end|off|reset) __attr=$__end;;
    bold)          __attr='\033[1m';;
    dim)           __attr='\033[2m';;
    italic)        __attr='\033[4m';;
    blink)         __attr='\033[5m';;
    invert)        __attr='\033[7m';;
    hidden)        __attr='\033[8m';;
  esac
  if test "$2"; then
    echo -en "$__attr$2$__end"
  else
    echo -en "$__attr"
  fi
}
function a {
  attribute "$@"
}

## Elements
function br {
  echo -e "\n\r"
}

function tab {
  echo -e "\t"
}

function indent {
  local __len=4
  if test "$1"; then
    if [[ $1 =~ $re ]] ; then
      __len=$1
    fi
  fi
  while [ $__len -gt 0 ]; do
    echo -n " "
     __len=$(( $__len - 1 ))
  done
}
function i {
  indent "$@"
}

function hr {
  local __len=60
  local __char='-'
  if ! test "$2"; then
    re='^[0-9]+$'
    if [[ $1 =~ $re ]] ; then
      __len=$1
    elif test "$1"; then
      __char=$1
    fi
  else
    __len=$2
    __char=$1
  fi
  while [ $__len -gt 0 ]; do
    echo -n "$__char"
     __len=$(( $__len - 1 ))
  done
}

# Icons
##
declare -A entities=()
entities[quot]='\u0022'
entities[amp]='\u0026'
entities[lt]='\u003C'
entities[gt]='\u003E'
entities[nbsp]=' '
entities[pi]='\u03C0'
entities[ndash]='\u2013'
entities[mdash]='\u2014'
entities[lsquo]='\u2018'
entities[rsquo]='\u2019'
entities[sbquo]='\u201A'
entities[ldquo]='\u201C'
entities[rdquo]='\u201D'
entities[dagger]='\u2020'
entities[bullet]='\u2022'
entities[lsaquo]='\u2039'
entities[rasquo]='\u203A'
entities[oline]='\u203E'
entities[frasl]='\u2044'
entities[euro]='\u20AC'
entities[larr]='\u2190'
entities[uarr]='\u2191'
entities[rarr]='\u2192'
entities[darr]='\u2193'
entities[harr]='\u2194'
entities[cararr]='\u21B5'
entities[lArr]='\u21D0'
entities[uArr]='\u21D1'
entities[rArr]='\u21D2'
entities[dArr]='\u21D3'
entities[hArr]='\u21D4'
entities[empty]='\u2205'
entities[sum]='\u2211'
entities[minus]='\u2212'
entities[spades]='\u2660'
entities[clubs]='\u2663'
entities[hearts]='\u2665'
entities[diams]='\u2666'
function _entity_list {
  local w=0
  for i in "${!entities[@]}"; do
    if [ $w -lt 3 ]; then
      printf "  %07s:  %s   " "$i" "$( echo -e "${entities[$i]}")"
      w="$(expr $w + 1)"
    else
      printf "  %07s:  %s   \n" "$i" "$( echo -e "${entities[$i]}")"
      w=0
    fi
  done
  echo " "
}
function entity {
  echo -ne "${entities[$1]}"
}
function icon {
  local i='';
  case "$1" in
    check|checkmark)       i='\xE2\x9C\x93';;
    X|x|xmark)             i='\xE2\x9C\x98';;
    '<3'|heart)            i='\xE2\x9D\xA4';;
    sun)                   i='\xE2\x98\x80';;
    '*'|star)              i='\xE2\x98\x85';;
    darkstar)              i='\xE2\x98\x86';;
    umbrella)              i='\xE2\x98\x82';;
    flag)                  i='\xE2\x9A\x91';;
    snow|snowflake)        i='\xE2\x9D\x84';;
    music)                 i='\xE2\x99\xAB';;
    scissors)              i='\xE2\x9C\x82';;
    tm|trademark)          i='\xE2\x84\xA2';;
    copyright)             i='\xC2\xA9';;
    apple)                 i='\xEF\xA3\xBF';;
    skull|bones)           i='\xE2\x98\xA0';;
    ':-)'|':)'|smile|face) i='\xE2\x98\xBA';;
    *)
      entity $1; return 0;;
  esac
  echo -ne "$i";
}

#SHML:END

if echo "$*" | grep "\-\-version" > /dev/null; then
  echo "$0 version $SHUNT_VERSION"
  exit 0
fi

__usage() {
cat << EOF
Usage: $0 <test files>

Options:
--plain    Disable colors and icons.
--quiet    Do not print error messages.
--verbose  Display success messages.
--version  Display version information.
--help     Display this message.

EOF
exit 0
}

options="$@"

# Usage - TODO: iterate of arguments
##
if echo "$options" | grep "\-h" > /dev/null; then # also matches '--help'
  __usage
fi

# Style
##
__no_style=false
if echo "$options" | grep "\-\-plain" > /dev/null; then
  __no_style=true
  options="$(echo "$options" | sed 's/--plain//')"
fi

# quiet
##
__quiet=false
if echo "$options" | grep "\-\-quiet" > /dev/null; then
  __quiet=true
  options="$(echo "$options" | sed 's/--quiet//')"
fi

# Verbose
##
__verbose=false
if echo "$options" | grep "\-\-verbose" > /dev/null; then
  __verbose=true
  options="$(echo "$options" | sed 's/--verbose//')"
fi

# Files
##
if ! test "$options"; then __usage; fi

# Before/After function handling
##
function __ensure_handlers {
  _="$( { type before; } 2>&1 )"
  if [ "$?" -ne "0" ]; then function before { true; }; fi
  _="$( { type after; } 2>&1 )"
  if [ "$?" -ne "0" ]; then function after { true; }; fi
}

# Progress Variables
__total=0; __passed=0; __failed=0; __failures=""; __successes=""; __current=""; __last=""
################################################################################
# Assertions
################################################################################
function assert {
  local cmd=$1
  local msg="[assert] $2"
  err="$( { $cmd; } 2>&1 )"
  process "$?" "$msg" "$cmd" "$err"
  unset err # status is unavailable when err is set to local
}

function refute {
  local cmd=$1
  local msg="[refute] $2"
  _="$( { $cmd; } 2>&1)"
  [ "$?" -ne "0" ]
  process "$?" "$msg" "$cmd" "exit status is zero"
}

function assert_equal {
  local one=$1
  local two=$2
  local msg="[assert_equal] $3"
  [ "$one" = "$two" ]
  process "$?" "$msg" "" "'$one' does not equal '$two'"
}

function refute_equal {
  local one=$1
  local two=$2
  local msg="[refute_equal] $3"
  [ "$one" != "$two" ]
  process "$?" "$msg" "" "'$one' equals '$two'"
}

function assert_numeq {
  local one=$1
  local two=$2
  local msg="[assert_numeq] $3"
  [ "$one" -eq "$two" ]
  process "$?" "$msg" "" "'$one' does not equal '$two'"
}

function refute_numeq {
  local one=$1
  local two=$2
  local msg="[refute_numeq] $3"
  [ "$one" -ne "$two" ]
  process "$?" "$msg" "" "'$one' equals '$two'"
}

function assert_grep {
  local cmd=$1
  local inc=$2
  local msg="[assert_grep] $3"
  out="$( { $cmd; } 2>&1 )"
  _="$( { echo $out | grep $inc; } 2>&1 )"
  process "$?" "$msg" "$cmd" "'$out' does not include '$inc'"
  unset out
}

function refute_grep {
  local cmd=$1
  local inc=$2
  local msg="[refute_grep] $3"
  out="$( { $cmd; } 2>&1 )"
  _="$( { echo $out | grep -v $inc; } 2>&1 )"
  process "$?" "$msg" "$cmd" "'$out' includes '$inc'"
  unset out
}

function assert_file {
  local file=$1
  local msg="[assert_file] $2"
  _="$( { test -f $file; } 2>&1 )"
  process "$?" "$msg" "" "file '$file' does not exist"
}

function refute_file {
  local file=$1
  local msg="[refute_file] $2"
  _="$( { test -f $file; } 2>&1 )"
  [ "$?" -ne "0" ]
  process "$?" "$msg" "" "file '$file' exists"
}

function assert_dir {
  local dir=$1
  local msg="[assert_dir] $2"
  _="$( { test -d $dir; } 2>&1 )"
  process "$?" "$msg" "" "directory '$dir' does not exist"
}

function refute_dir {
  local dir=$1
  local msg="[refute_dir] $2"
  _="$( { test -d $dir; } 2>&1 )"
  [ "$?" -ne "0" ]
  process "$?" "$msg" "" "directory '$dir' exists"
}

################################################################################
# Utils
################################################################################
function process {
  local status=$1
  local msg=$2
  local cmd=$3
  local err=$4
  if [ "$status" -eq "0" ]; then
    __do_pass "$msg"
  else
    __do_fail "$msg" "$cmd" "$err"
  fi
}

function __do_check {
  if $__no_style; then
    echo -ne "."
  else
    echo -ne "$(color green "$(icon check)")"
  fi
}

function __do_x {
  if $__no_style; then
    echo -ne "x"
  else
    echo -ne "$(color red "$(icon x)")"
  fi
}

function __do_pass {
  local msg=$1
  __total=$(expr $__total + 1)
  __passed=$(expr $__passed + 1)
  if $__verbose; then
    echo "$__total. $(__do_color green "$msg passed")"
  else
    __do_check
  fi
}

function __do_fail {
  local msg=$1
  local cmd=$2
  local err=$3
  __total=$(expr $__total + 1)
  __failed=$(expr $__failed + 1)

  if $__verbose; then
    echo "$__total. $(__do_color red "$msg failed")"
  else
    __do_x
  fi

  if [ "$__last" != "$__current" ]; then
    __failures+="$(br)$__current$(br)"
    __last=$__current
  fi

  __failures+="$(__do_color red "$__total. $msg")$(br)"
  if test "$err" && ! $__quiet; then
    if test "$cmd"; then
      __failures+="$(__do_color yellow "$(i 2)'$cmd' failed with:$(br)$(i 2)")"
    fi
    __failures+="$(__do_color yellow "$(i 2) $err$(br)$(br)")"
  fi
}

function __do_color {
  if $__no_style; then
    echo "$2"
  else
    echo "$(color $1 "$2")"
  fi
}

function __failures {
  if [ "$__failed" -ne 0 ] && ! $__quiet; then
    echo "$(br)$(br)Failures:$(br)$(hr)$(br)$__failures"
  else
    echo " "
  fi
}

function __finish {
  __failures
  if $__verbose; then echo -n "$(hr)"; fi
  if $__quiet; then echo " "; fi
  echo "$(__do_color yellow "$(br)Total: $(expr $__passed + $__failed)")  $(__do_color green "Passed: $__passed")  $(__do_color red "Failed: $__failed")  $(__do_color blue "Duration: ${SECONDS} Seconds")$(br)"
}

function __reset {
  unset before
  unset after
}

function __shunt {
  __ensure_handlers
  before
  run_tests
  after
}

################################################################################
# Make is so.
################################################################################

echo "$(basename -- $0) $@"
if $__verbose; then echo "$(hr '=')"; fi
echo " "

# In an edge case, a test changes directory. It needs to be changed
# back after the test.
#######
here="$(pwd)"

for __current in $options; do
  __reset
  source $__current
  if $__verbose; then echo "$(br)$__current$(br)$(hr)$(br)"; fi
  __shunt
  cd $here
done

__finish
exit $__failed

# vim: ft=sh:


install.sh
#!/usr/bin/env bash

set -eET

# Options
##
__usage()
{
  cat << EOF
Usage: $0 [BRANCH] [INSTALL METHOD]

 INSTALL METHOD:
    global   Install globally.
    user     Install to user. (default)
    local    Install to current directory.

 OPTIONS:
    --help     Show this message.

EOF
exit 0
}

_sudo=""
if test "$*"; then
  options="$*"
  if echo "$options" | grep "\-\-help" > /dev/null; then
    __usage
  fi
  if echo "$options" | grep "global" > /dev/null; then
    install_method=global
    options="$(echo "$options" | sed 's/global//')"
    install_path="/usr/local/bin"
    if (( UID != 0 )); then
      _sudo="sudo "
    fi
  elif echo "$options" | grep "local" > /dev/null; then
    install_method=local
    install_path="."
    options="$(echo "$options" | sed 's/local//')"
  elif echo "$options" | grep "user" > /dev/null; then
    install_method=user
    install_path="$HOME/.bin"
    options="$(echo "$options" | sed 's/user//')"
  fi
fi

# setup
##
options=$( echo $options )
test "$options" && version=$options
test "$version" || version="latest"
source="https://raw.github.com/odb/shunt/$version/shunt.sh"
target="shunt.sh"
execln="shunt"

# display banner
##
echo "Installing shunt"
echo "------------------"
echo " "

# check superuser if method not specified
##
if ! test "$install_method"; then
  install_method="user"
  install_path="$HOME/.bin"
  if (( UID == 0 )); then
    install_method="global"
    install_path="/usr/local/bin"
  fi
fi

# ensure install directory
##
test -d $install_path || $_sudo mkdir -p $install_path

# fetch latest
##
cd $install_path
if test -f $target; then
  echo "> Backing up previous version:"
  $_sudo mv -v $target $target.bak
  echo " "
fi
echo "> Downloading $source:"
$_sudo curl -O $source
$_sudo chmod 755 $install_path/$target

# create executable symlink
##
if [ "$install_method" != "local" ]; then
  test -L $execln || $_sudo ln -s $target $execln
  echo "> Install path: $install_path/$execln"
else
  echo "> Install path: $install_path/$target"
fi

# update path
##

export_string="\nexport PATH=$install_path:\$PATH # Add shunt to PATH"
report=false
if [ "$install_method" = "user" ]; then
  if test -f $HOME/.zshrc; then
    if ! grep "$install_path" $HOME/.zshrc 2>&1 > /dev/null; then
      echo -e "$export_string" >> $HOME/.zshrc
    fi
  elif test -f $HOME/.profile; then
    if ! grep "$install_path" $HOME/.profile 2>&1 > /dev/null; then
      echo -e "$export_string" >> $HOME/.profile
      report=true
    fi
  elif test -f $HOME/.bashrc; then
    if ! grep "$install_path" $HOME/.bashrc 2>&1 > /dev/null; then
      echo -e "$export_string" >> $HOME/.bashrc
      report=true
    fi
  else
    echo " "
    echo "WARNING: Please add $install_path to your PATH."
  fi
else
  report=false
fi

if $report; then
  echo " "
  echo "NOTE: Please log out and log back in to ensure that 'shunt' is available to your shell."
fi

echo " "
echo "DONE"
# vim: ft=sh:


example.sh
#!/usr/bin/env bash

# Setup
##
touch /tmp/cliunit.file
mkdir -p /tmp/cliunit.dir

# Test Examples
##

before_check=false
function before {
  before_check=true
}

function after {
  echo "If you see this, after worked."
}

function run_tests {

  assert "true"  "should assert truth"
  refute "false" "should refute truth"

  assert_equal "true" "true" "should assert equality"
  refute_equal "true" "false" "should refute equality"

  assert_numeq "1111" "1111" "should assert numerical equality"
  refute_numeq "1111" "1112" "should refute numerical equality"

  assert_grep "echo foobar" "foobar" "should assert via grep"
  refute_grep "echo foobar" "notfoobar" "should refute via grep"

  assert_file "/tmp/cliunit.file" "should assert file existence"
  refute_file "/tmp/cliunit.bad"  "should refute file existence"

  assert_dir "/tmp/cliunit.dir" "should assert directory existence"
  refute_dir "/tmp/cliunit.bad" "should refute directory existence"

  sleep 0.5 # showing Duration > 0

  assert "false" "should fail, to show failure output"

  assert "$before_check" "before failed"
}
source ./CLIunit.sh

# vim: ft=sh:
